src/main/java/org/hyperledger/fabric/sdk/security/CryptoPrimitives.java

```diff
@@ -1007,4 +984,173 @@ public byte[] certificateToDER(String certificatePEM) {
+        public int hashCode() {
+            return hashCode;
+        }
+
+    }
+
+
+    /** A cached X.509 certificate. */
+    private class CertValue {
+        /** Link back to the cache key. */
+        final CertKey key;
+
+        /** Indicator if the X.509 certificate is currently valid. Note: it may become valid at some point in the future. */
+        boolean isValid;
+
+        /** The next time this certificate will be checked, in milliseconds since the epoch. */
+        long nextCheckTime;
+
+        /** The certificate derived from the PEM data. */
+        X509Certificate x509;
+
+
+        /**
+         * New instance.
+         *
+         * @param key the cache key
+         */
+        CertValue(CertKey key) {
+            this.key = key;
+            x509 = null;
+            isValid = false;
+            nextCheckTime = Long.MIN_VALUE;
+        }
+
+
+        /**
+         * Get the certificate if it is valid.
+         *
+         * @return the certificate, or null
+         *
+         * @throws CryptoException if the certificate cannot be parsed.
+         */
+        public synchronized X509Certificate getValid() throws CryptoException {
+            return isValid() ? getX509() : null;
+        }
+
+
+        /**
+         * Get the X.509 certificate.
+         *
+         * @return the certificate
+         *
+         * @throws CryptoException if the PEM data cannot be parsed.
+         */
+        public synchronized X509Certificate getX509() throws CryptoException {
+            if (x509 == null) {
+                // If the security providers change then something that could not be parsed could become parsable.
+                try {
+                    x509 = getX509CertificateInternal(key.pemData);
+                } catch (CryptoException e) {
+                    isValid = false;
+                    nextCheckTime = System.currentTimeMillis() + X509_RECHECK_MILLIS;
+                    logger.error("Unable to recover X.509 certificate from provided binary data", e);
+                    throw e;
+                }
+            }
+            return x509;
+        }
+
+
+        /**
+         * Is the X.509 certificate currently valid?
+         *
+         * @return true if the certificate is valid
+         *
+         * @throws CryptoException if the certificate cannot be validated
+         */
+        public synchronized boolean isValid() throws CryptoException {
+            // Only re-check if a reasonable amount of time has passed.
+            long now = System.currentTimeMillis();
+            if (nextCheckTime >= now) {
+                return isValid;
+            }
+
+            nextCheckTime = now + X509_RECHECK_MILLIS;
+            isValid = validateCertificate(getX509());
+            return isValid;
+        }
+
+    }
+
+
+    private final LoadingCache<CertKey, CertValue> x509Cache = CacheBuilder.newBuilder().maximumSize(1000).build(new CacheLoader<CertKey, CertValue>() {
```

Comment:
I wonder if an `expireAfterAccess` setting would also be beneficial to ensure the cache does't collect stale or very rarely used values over long execution times. What do you think?

Reply:
The `expireAfterAccess` setting doesn't do what one imagines, unless you also have a timer thread periodically invoking the cache's `cleanUp` method. Entries that are past their expiry time are kept in memory until some process touches them. There is no automatic clean up of Guava caches.

Using `expireAfterWrite` could have been used to simplify the implementation at the cost of re-parsing and re-checking the certificate periodically where currently we just need to do the re-check to spot certificates reaching their end of life. Less code, but also less performance, and I've written the code now :) 

Is there a mechanism in the SDK for specifying low level configuration? A configurable cache size instead of a fixed one would be a nice-to-have in some circumstances.

For the time being, I would suggest just leaving this as it is.

Reply:
There are a few values obtained from config at the top of CryptoPrimitives. You could use the same pattern quite easily, just adding another property with suitable default to `org.hyperledger.fabric.sdk.helper.Config`.